lambdas/s3hash/src/t4_lambda_s3hash/__init__.py

```diff
@@ -0,0 +1,388 @@
+
+
+async def get_existing_checksum(
+    location: S3ObjectSource,
+    parts: T.Sequence[PartDef],
+) -> T.Optional[Checksum]:
+    # XXX
+    # get object attributes and see if part sizes and checksum algo matches our standard
+    # return the existing checksum if it's compliant
+    # otherwise return None
+    return None
+
+
+def hash_parts(parts: T.Sequence[bytes]) -> bytes:
+    return hashlib.sha256(b"".join(parts)).digest()
+
+
+class PartDef(pydantic.BaseModel):
+    part_number: int
+    range: T.Optional[T.Tuple[int, int]]
+
+    @property
+    def boto_args(self):
+        args: T.Dict[str, T.Any] = {"PartNumber": self.part_number}
+        if self.range:
+            args["CopySourceRange"] = f"bytes={self.range[0]}-{self.range[1]}"
+        return args
+
+    @classmethod
+    def for_range(cls, part_number: int, offset: int, size: int):
+        return cls(
+            part_number=part_number,
+            range=(offset, offset + size - 1),
+        )
+
+
+PARTS_SINGLE = [PartDef(part_number=1, range=None)]
+
+
+async def get_parts_for_location(location: S3ObjectSource) -> T.List[PartDef]:
+    total_size = (await S3.get().head_object(**location.boto_args))["ContentLength"]
+    part_size = get_part_size(total_size)
+
+    # single-part upload
+    if part_size is None:
+        return PARTS_SINGLE
+
+    # multipart upload
+    offset = 0
+    part_number = 1
+    parts = []
+    while offset < total_size:
+        actual_part_size = min(part_size, total_size - offset)
+        parts.append(
+            PartDef.for_range(
+                part_number=part_number,
+                offset=offset,
+                size=actual_part_size,
+            )
+        )
+        offset += actual_part_size
+        part_number += 1
+
+    return parts
+
+
+@tenacity.retry
+async def upload_part(mpu: MPURef, src: S3ObjectSource, part: PartDef) -> bytes:
+    res = await S3.get().upload_part_copy(
+        **mpu.boto_args,
+        **part.boto_args,
+        CopySource=src.boto_args,
+    )
+    cs = res["CopyPartResult"].get("ChecksumSHA256")  # base64-encoded
+    assert cs is not None
+    return base64.b64decode(cs)
+
+
+async def compute_part_checksums(
+    mpu: MPURef,
+    location: S3ObjectSource,
+    parts: T.Sequence[PartDef],
+) -> T.Tuple[T.List[bytes], T.Any]:
+    uploads = [upload_part(mpu, location, p) for p in parts]
+    checksums: T.List[bytes] = await asyncio.gather(*uploads)
+    return checksums, upload_part.retry.statistics
+
+
+async def _create_mpu(target: S3ObjectDestination) -> MPURef:
+    upload_data = await S3.get().create_multipart_upload(
+        **target.boto_args,
+        ChecksumAlgorithm="SHA256",
+    )
+    return MPURef(bucket=target.bucket, key=target.key, id=upload_data["UploadId"])
+
+
+def get_mpu_dsts(src: S3ObjectSource, target: T.Optional[S3ObjectDestination]) -> T.List[S3ObjectDestination]:
+    dsts = [
+        S3ObjectDestination(bucket=src.bucket, key=SCRATCH_KEY_PER_BUCKET),
```

Comment:
That's a bit risky optimization. It looks like we can't strongly guarantee we call `abort_multipart_upload()` and customers can have buckets without lifecycle policy that aborts uploads automatically (though I think it's advised by AWS), so unfinished uploads will stay forever burning some money. I *think* most customers have their bucket in the same region, it might be wiser to remove this optimization.

Reply:
ok, that makes sense